docs/benchmarking.md

```diff
@@ -4,28 +4,137 @@ toc: true
 mathjax: true
 ---
 
-In this document we discuss your options when it comes to benchmarking code written using Cryptimeleon libraries as well as group operation counting provided by Cryptimeleon Math.
+In this document we discuss your options when it comes to benchmarking code written using Cryptimeleon libraries.
 
-# Runtime Benchmarking
+There are many different kinds of possible performance metrics.
+These include runtime in the form of CPU cycles or CPU time, memory usage, and network usage.
+Furthermore, one may want to collect hardware-independent information such as the number of group operations or pairings.
+Both of these types of metrics have their use-cases.
+Counting group operations and pairings has the advantage of being hardware-independent.
+To the layperson and potential user, applied metrics such as CPU time or memory usage can be more meaningful as they demonstrate practicality better than the more abstract group operations metrics.
 
-## Lazy Eval
+# Collecting Applied Metrics
+
+Applied metrics, such as runtime or memory usage, can be collected using any existing Java benchmark framework.
+An example of such a framework is the Java Microbenchmarking Harness (JMH). 
+It allows for very accurate measurements and integrates with many existing profilers.
+Due to these existing options, we have decided against implementing any such capabilities ourselves.
+Therefore, you will need to use one of these existing benchmark frameworks.
+
+## Problems That Can Falsify Your Benchmarks
+
+We now want to look at some potential problems when creating a benchmark.
+We will illustrate these problems using the example of benchmarking the verification algorithm of a signature scheme.
+
+### Lazy Evaluation
 
 While useful for automatic optimization, the [lazy evaluation]({% link docs/lazy-eval.md %}) features of Math can create some benchmarking problems if not handled correctly.
+To benchmark your verification algorithm, you will need valid signatures, and these signatures will be provided by executing the signing algorithm.
+Inside the signing algorithm group operations might be executed.
+By default, group operations in Cryptimeleon Math are evaluated lazily, i.e. deferred until needed.
+Therefore, group operations done during signing will only be executed once the result is needed during verification of your signature.
+This will increase the runtime of your verification algorithm and therefore falsify the results.
+By serializing the `Signature` object using the `getRepresentation()` method *before* running verification, you can force all remaining computations involving fields of the `Signature` object to be completed in a blocking manner.
+
+### Mutability
+
+Another problem is the mutability of the `Signature` object.
```

Comment:
I don't know whether there's a better word for it. Like ... the value represented by the object never changes. It's not _mutable_ in this sense. There are just some helper values that may be added that don't influence the value, just performance characteristics.